// @aws-sdk/client-neptune/dist-es/models/models_0.js
import { SENSITIVE_STRING } from "@smithy/smithy-client";
import { NeptuneServiceException as __BaseException } from "./NeptuneServiceException";
export class DBClusterNotFoundFault extends __BaseException {
    name = "DBClusterNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBClusterNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBClusterNotFoundFault.prototype);
    }
}
export class DBClusterRoleAlreadyExistsFault extends __BaseException {
    name = "DBClusterRoleAlreadyExistsFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBClusterRoleAlreadyExistsFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBClusterRoleAlreadyExistsFault.prototype);
    }
}
export class DBClusterRoleQuotaExceededFault extends __BaseException {
    name = "DBClusterRoleQuotaExceededFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBClusterRoleQuotaExceededFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBClusterRoleQuotaExceededFault.prototype);
    }
}
export class InvalidDBClusterStateFault extends __BaseException {
    name = "InvalidDBClusterStateFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidDBClusterStateFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidDBClusterStateFault.prototype);
    }
}
export class SourceNotFoundFault extends __BaseException {
    name = "SourceNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "SourceNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, SourceNotFoundFault.prototype);
    }
}
export class SubscriptionNotFoundFault extends __BaseException {
    name = "SubscriptionNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "SubscriptionNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, SubscriptionNotFoundFault.prototype);
    }
}
export class DBInstanceNotFoundFault extends __BaseException {
    name = "DBInstanceNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBInstanceNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBInstanceNotFoundFault.prototype);
    }
}
export class DBSnapshotNotFoundFault extends __BaseException {
    name = "DBSnapshotNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBSnapshotNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBSnapshotNotFoundFault.prototype);
    }
}
export class ResourceNotFoundFault extends __BaseException {
    name = "ResourceNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ResourceNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceNotFoundFault.prototype);
    }
}
export class DBParameterGroupAlreadyExistsFault extends __BaseException {
    name = "DBParameterGroupAlreadyExistsFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBParameterGroupAlreadyExistsFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBParameterGroupAlreadyExistsFault.prototype);
    }
}
export class DBParameterGroupNotFoundFault extends __BaseException {
    name = "DBParameterGroupNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBParameterGroupNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBParameterGroupNotFoundFault.prototype);
    }
}
export class DBParameterGroupQuotaExceededFault extends __BaseException {
    name = "DBParameterGroupQuotaExceededFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBParameterGroupQuotaExceededFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBParameterGroupQuotaExceededFault.prototype);
    }
}
export class DBClusterSnapshotAlreadyExistsFault extends __BaseException {
    name = "DBClusterSnapshotAlreadyExistsFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBClusterSnapshotAlreadyExistsFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBClusterSnapshotAlreadyExistsFault.prototype);
    }
}
export class DBClusterSnapshotNotFoundFault extends __BaseException {
    name = "DBClusterSnapshotNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBClusterSnapshotNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBClusterSnapshotNotFoundFault.prototype);
    }
}
export class InvalidDBClusterSnapshotStateFault extends __BaseException {
    name = "InvalidDBClusterSnapshotStateFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidDBClusterSnapshotStateFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidDBClusterSnapshotStateFault.prototype);
    }
}
export class KMSKeyNotAccessibleFault extends __BaseException {
    name = "KMSKeyNotAccessibleFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "KMSKeyNotAccessibleFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, KMSKeyNotAccessibleFault.prototype);
    }
}
export class SnapshotQuotaExceededFault extends __BaseException {
    name = "SnapshotQuotaExceededFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "SnapshotQuotaExceededFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, SnapshotQuotaExceededFault.prototype);
    }
}
export class DBClusterAlreadyExistsFault extends __BaseException {
    name = "DBClusterAlreadyExistsFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBClusterAlreadyExistsFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBClusterAlreadyExistsFault.prototype);
    }
}
export class DBClusterParameterGroupNotFoundFault extends __BaseException {
    name = "DBClusterParameterGroupNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBClusterParameterGroupNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBClusterParameterGroupNotFoundFault.prototype);
    }
}
export class DBClusterQuotaExceededFault extends __BaseException {
    name = "DBClusterQuotaExceededFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBClusterQuotaExceededFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBClusterQuotaExceededFault.prototype);
    }
}
export class DBSubnetGroupDoesNotCoverEnoughAZs extends __BaseException {
    name = "DBSubnetGroupDoesNotCoverEnoughAZs";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBSubnetGroupDoesNotCoverEnoughAZs",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBSubnetGroupDoesNotCoverEnoughAZs.prototype);
    }
}
export class DBSubnetGroupNotFoundFault extends __BaseException {
    name = "DBSubnetGroupNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBSubnetGroupNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBSubnetGroupNotFoundFault.prototype);
    }
}
export class GlobalClusterNotFoundFault extends __BaseException {
    name = "GlobalClusterNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "GlobalClusterNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, GlobalClusterNotFoundFault.prototype);
    }
}
export class InsufficientStorageClusterCapacityFault extends __BaseException {
    name = "InsufficientStorageClusterCapacityFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InsufficientStorageClusterCapacityFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InsufficientStorageClusterCapacityFault.prototype);
    }
}
export class InvalidDBInstanceStateFault extends __BaseException {
    name = "InvalidDBInstanceStateFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidDBInstanceStateFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidDBInstanceStateFault.prototype);
    }
}
export class InvalidDBSubnetGroupStateFault extends __BaseException {
    name = "InvalidDBSubnetGroupStateFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidDBSubnetGroupStateFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidDBSubnetGroupStateFault.prototype);
    }
}
export class InvalidGlobalClusterStateFault extends __BaseException {
    name = "InvalidGlobalClusterStateFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidGlobalClusterStateFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidGlobalClusterStateFault.prototype);
    }
}
export class InvalidSubnet extends __BaseException {
    name = "InvalidSubnet";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidSubnet",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidSubnet.prototype);
    }
}
export class InvalidVPCNetworkStateFault extends __BaseException {
    name = "InvalidVPCNetworkStateFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidVPCNetworkStateFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidVPCNetworkStateFault.prototype);
    }
}
export class StorageQuotaExceededFault extends __BaseException {
    name = "StorageQuotaExceededFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "StorageQuotaExceededFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, StorageQuotaExceededFault.prototype);
    }
}
export class DBClusterEndpointAlreadyExistsFault extends __BaseException {
    name = "DBClusterEndpointAlreadyExistsFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBClusterEndpointAlreadyExistsFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBClusterEndpointAlreadyExistsFault.prototype);
    }
}
export class DBClusterEndpointQuotaExceededFault extends __BaseException {
    name = "DBClusterEndpointQuotaExceededFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBClusterEndpointQuotaExceededFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBClusterEndpointQuotaExceededFault.prototype);
    }
}
export class AuthorizationNotFoundFault extends __BaseException {
    name = "AuthorizationNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "AuthorizationNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AuthorizationNotFoundFault.prototype);
    }
}
export class DBInstanceAlreadyExistsFault extends __BaseException {
    name = "DBInstanceAlreadyExistsFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBInstanceAlreadyExistsFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBInstanceAlreadyExistsFault.prototype);
    }
}
export class DBSecurityGroupNotFoundFault extends __BaseException {
    name = "DBSecurityGroupNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBSecurityGroupNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBSecurityGroupNotFoundFault.prototype);
    }
}
export class DomainNotFoundFault extends __BaseException {
    name = "DomainNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DomainNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DomainNotFoundFault.prototype);
    }
}
export class InstanceQuotaExceededFault extends __BaseException {
    name = "InstanceQuotaExceededFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InstanceQuotaExceededFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InstanceQuotaExceededFault.prototype);
    }
}
export class InsufficientDBInstanceCapacityFault extends __BaseException {
    name = "InsufficientDBInstanceCapacityFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InsufficientDBInstanceCapacityFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InsufficientDBInstanceCapacityFault.prototype);
    }
}
export class OptionGroupNotFoundFault extends __BaseException {
    name = "OptionGroupNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "OptionGroupNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, OptionGroupNotFoundFault.prototype);
    }
}
export class ProvisionedIopsNotAvailableInAZFault extends __BaseException {
    name = "ProvisionedIopsNotAvailableInAZFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ProvisionedIopsNotAvailableInAZFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ProvisionedIopsNotAvailableInAZFault.prototype);
    }
}
export class StorageTypeNotSupportedFault extends __BaseException {
    name = "StorageTypeNotSupportedFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "StorageTypeNotSupportedFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, StorageTypeNotSupportedFault.prototype);
    }
}
export class DBSubnetGroupAlreadyExistsFault extends __BaseException {
    name = "DBSubnetGroupAlreadyExistsFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBSubnetGroupAlreadyExistsFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBSubnetGroupAlreadyExistsFault.prototype);
    }
}
export class DBSubnetGroupQuotaExceededFault extends __BaseException {
    name = "DBSubnetGroupQuotaExceededFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBSubnetGroupQuotaExceededFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBSubnetGroupQuotaExceededFault.prototype);
    }
}
export class DBSubnetQuotaExceededFault extends __BaseException {
    name = "DBSubnetQuotaExceededFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBSubnetQuotaExceededFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBSubnetQuotaExceededFault.prototype);
    }
}
export class EventSubscriptionQuotaExceededFault extends __BaseException {
    name = "EventSubscriptionQuotaExceededFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "EventSubscriptionQuotaExceededFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, EventSubscriptionQuotaExceededFault.prototype);
    }
}
export class SNSInvalidTopicFault extends __BaseException {
    name = "SNSInvalidTopicFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "SNSInvalidTopicFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, SNSInvalidTopicFault.prototype);
    }
}
export class SNSNoAuthorizationFault extends __BaseException {
    name = "SNSNoAuthorizationFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "SNSNoAuthorizationFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, SNSNoAuthorizationFault.prototype);
    }
}
export class SNSTopicArnNotFoundFault extends __BaseException {
    name = "SNSTopicArnNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "SNSTopicArnNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, SNSTopicArnNotFoundFault.prototype);
    }
}
export class SubscriptionAlreadyExistFault extends __BaseException {
    name = "SubscriptionAlreadyExistFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "SubscriptionAlreadyExistFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, SubscriptionAlreadyExistFault.prototype);
    }
}
export class SubscriptionCategoryNotFoundFault extends __BaseException {
    name = "SubscriptionCategoryNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "SubscriptionCategoryNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, SubscriptionCategoryNotFoundFault.prototype);
    }
}
export const FailoverStatus = {
    CANCELLING: "cancelling",
    FAILING_OVER: "failing-over",
    PENDING: "pending",
};
export class GlobalClusterAlreadyExistsFault extends __BaseException {
    name = "GlobalClusterAlreadyExistsFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "GlobalClusterAlreadyExistsFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, GlobalClusterAlreadyExistsFault.prototype);
    }
}
export class GlobalClusterQuotaExceededFault extends __BaseException {
    name = "GlobalClusterQuotaExceededFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "GlobalClusterQuotaExceededFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, GlobalClusterQuotaExceededFault.prototype);
    }
}
export class DBClusterEndpointNotFoundFault extends __BaseException {
    name = "DBClusterEndpointNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBClusterEndpointNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBClusterEndpointNotFoundFault.prototype);
    }
}
export class InvalidDBClusterEndpointStateFault extends __BaseException {
    name = "InvalidDBClusterEndpointStateFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidDBClusterEndpointStateFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidDBClusterEndpointStateFault.prototype);
    }
}
export class InvalidDBParameterGroupStateFault extends __BaseException {
    name = "InvalidDBParameterGroupStateFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidDBParameterGroupStateFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidDBParameterGroupStateFault.prototype);
    }
}
export class DBSnapshotAlreadyExistsFault extends __BaseException {
    name = "DBSnapshotAlreadyExistsFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBSnapshotAlreadyExistsFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBSnapshotAlreadyExistsFault.prototype);
    }
}
export class InvalidDBSubnetStateFault extends __BaseException {
    name = "InvalidDBSubnetStateFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidDBSubnetStateFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidDBSubnetStateFault.prototype);
    }
}
export class InvalidEventSubscriptionStateFault extends __BaseException {
    name = "InvalidEventSubscriptionStateFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidEventSubscriptionStateFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidEventSubscriptionStateFault.prototype);
    }
}
export const ApplyMethod = {
    immediate: "immediate",
    pending_reboot: "pending-reboot",
};
export const SourceType = {
    db_cluster: "db-cluster",
    db_cluster_snapshot: "db-cluster-snapshot",
    db_instance: "db-instance",
    db_parameter_group: "db-parameter-group",
    db_security_group: "db-security-group",
    db_snapshot: "db-snapshot",
};
export class InvalidDBSecurityGroupStateFault extends __BaseException {
    name = "InvalidDBSecurityGroupStateFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidDBSecurityGroupStateFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidDBSecurityGroupStateFault.prototype);
    }
}
export class SharedSnapshotQuotaExceededFault extends __BaseException {
    name = "SharedSnapshotQuotaExceededFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "SharedSnapshotQuotaExceededFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, SharedSnapshotQuotaExceededFault.prototype);
    }
}
export class CertificateNotFoundFault extends __BaseException {
    name = "CertificateNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "CertificateNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, CertificateNotFoundFault.prototype);
    }
}
export class DBUpgradeDependencyFailureFault extends __BaseException {
    name = "DBUpgradeDependencyFailureFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBUpgradeDependencyFailureFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBUpgradeDependencyFailureFault.prototype);
    }
}
export class SubnetAlreadyInUse extends __BaseException {
    name = "SubnetAlreadyInUse";
    $fault = "client";
    constructor(opts) {
        super({
            name: "SubnetAlreadyInUse",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, SubnetAlreadyInUse.prototype);
    }
}
export class DBClusterRoleNotFoundFault extends __BaseException {
    name = "DBClusterRoleNotFoundFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DBClusterRoleNotFoundFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DBClusterRoleNotFoundFault.prototype);
    }
}
export class InsufficientDBClusterCapacityFault extends __BaseException {
    name = "InsufficientDBClusterCapacityFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InsufficientDBClusterCapacityFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InsufficientDBClusterCapacityFault.prototype);
    }
}
export class InvalidDBSnapshotStateFault extends __BaseException {
    name = "InvalidDBSnapshotStateFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidDBSnapshotStateFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidDBSnapshotStateFault.prototype);
    }
}
export class InvalidRestoreFault extends __BaseException {
    name = "InvalidRestoreFault";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidRestoreFault",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidRestoreFault.prototype);
    }
}
export const CreateDBInstanceMessageFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.TdeCredentialPassword && { TdeCredentialPassword: SENSITIVE_STRING }),
});
export const ModifyDBInstanceMessageFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.TdeCredentialPassword && { TdeCredentialPassword: SENSITIVE_STRING }),
});
