// core-lib/lib/segmentHelper/definitions.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

// Titan Embedding
const INFERENCE_CONCURRENCY = 4;

// Segment known types
const KNOWN_TYPES = {
  TYPE_BLACKFRAMES: 'BlackFrames',
  TYPE_MONOCHROMEFRAMES: 'MonoColorFrames',
  TYPE_COLORBARS: 'ColorBars',
  TYPE_TECHNICAL_SLATE: 'TechnicalSlate',
  TYPE_COUNTDOWNCLOCK: 'CountdownClock',
  TYPE_IDENTS: 'Idents',
  TYPE_RATING: 'RatingsBand',
  TYPE_RECAP: 'Recap/Previously',
  TYPE_INTRO: 'Intro',
  TYPE_TITLE: 'Title',
  TYPE_MAINPROGRAM: 'MainProgram',
  TYPE_PROGRAMME: 'Programme',
  TYPE_OPENINGCREDITS: 'OpeningCredits',
  TYPE_ENDCREDITS: 'EndCredits',
  TYPE_NEXTEPISODE_CREDITS: 'NextEpisodeCredits',
  TYPE_TRANSITION: 'Transition',
  TYPE_POSTCREDITS_SCENE: 'PostCreditsScene',
  TYPE_TEXTLESSELEMENT: 'TextlessElement',
  TYPE_CREDITS: 'Credits', // both opening and end credits
  TYPE_UNDEFINED: 'Undefined',
};
const JSON_OUTPUTS = {
  JSON_FRAMEPROPS: 'frame_props.json',
  JSON_FRAMEEMBEDDINGS: 'frame_embeddings.json',
  JSON_FRAMESIMILARITY: 'frame_similarity.json',
  JSON_SHOTBOUNDARYFRAMES: 'shot_boundary_frames.json',
  JSON_FRAMESTOSHOTS: 'frames_to_shots.json',
  JSON_SHOTSTOSCENES: 'shots_to_scenes.json',
  JSON_SCENESTOCHAPTERS: 'scenes_to_chapters.json',
  JSON_SCENESEGMENTTYPE: 'scene_segment_type.json',
  JSON_SMPTE_ELEMENT: 'smpte_elements.json',
  // audio metadata (loudness)
  JSON_EBUR128: 'ebu_r128.json',
  JSON_LOUDNESSGROUP: 'loudness_group.json',
  // audio metadata (diarisations)
  JSON_DIARISATIONS: 'diarisations.json',
  JSON_SPEAKEREMBEDDINGS: 'speaker_embeddings.json',
  JSON_PAUSEINDIALOGUES: 'pause_in_dialogues.json',
  JSON_AUDIOSEGMENTS: 'audio_segments.json',
  // adbreak
  JSON_ADBREAK: 'adbreak.json',
  JSON_SCENE: 'scene.json',
  // face recognition
  JSON_FACEEMBEDDINGS: 'face_embeddings.json',
};
const FILTER_SETTINGS = {
  // Black frame threshold
  MAXPIXELTHRESHOLD: 0.2, // 0.02,
  MINCOVERAGEPERCENTAGE: 98, // 99.9998,
  // Shot/Scene threshold
  MINFRAMESIMILARITY: 0.75,
  MAXTIMEDISTANCE: 8000, // 8s
  MINSCENEDURATION: 4000, // 4s
  MAXSCENEDURATION: 180000, // 3min
  SCENESLIDINGWINDOW: 60000, // 1min
  MAXBOUNDARYFRAMES: 4,
  // Audio analysis
  GATEDLOUDNESS: -70.0,
  MAXLOUDNESSDISTANCE: 7,
  // loudness levels
  LUFS_ABSOLUTESILENT: -70.0,
  LUFS_VERYQUIET: -50.0,
  LUFS_QUIET: -30.0,
  LUFS_MODERATE: -10.0,
  LUFS_LOUD: 0.0,
  LUFS_VERYLOUD: 20.0,
  // label
  TAG_ABSOLUTESILENT: 'AbsoluteSilent',
  TAG_VERYQUIET: 'VeryQuiet',
  TAG_QUIET: 'Quiet',
  TAG_MODERATE: 'Moderate',
  TAG_LOUD: 'Loud',
  TAG_VERYLOUD: 'VeryLoud',
};


module.exports = {
  INFERENCE_CONCURRENCY,
  ...FILTER_SETTINGS,
  ...KNOWN_TYPES,
  ...JSON_OUTPUTS,
};
